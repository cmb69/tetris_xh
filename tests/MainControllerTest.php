<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Tetris_XH.
 *
 * Tetris_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Tetris_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tetris_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Tetris;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Tetris\Infra\FakeHighscoreService;
use Tetris\Infra\Jquery;
use Tetris\Infra\View;

class MainControllerTest extends TestCase
{
    private $sut;
    private $highscoreService;

    public function setUp(): void
    {
        $conf = XH_includeVar("./config/config.php", "plugin_cf")["tetris"];
        $text = XH_includeVar("./languages/en.php", "plugin_tx")["tetris"];
        $this->highscoreService = new FakeHighscoreService;
        $jquery = $this->createStub(Jquery::class);
        $view = new View("./views/", $text);
        $this->sut = new MainController("./plugins/tetris/", $conf, $this->highscoreService, $jquery, $view);
    }

    public function testRendersGame(): void
    {
        $response = ($this->sut)();
        Approvals::verifyHtml($response->output());
    }

    public function testReportsRequiredHighscore(): void
    {
        for ($i = 0; $i <10; $i++) {
            $this->highscoreService->enterHighscore("cmb", 4711);
        }
        $_GET = ["tetris_action" => "get_highscore"];
        $response = ($this->sut)();
        $this->assertTrue($response->terminated());
        $this->assertEquals("4711", $response->output());
    }

    public function testShowsHighscores(): void
    {
        $this->highscoreService->enterHighscore("cmb", 10000);
        $_GET = ["tetris_action" => "show_highscores"];
        $response = ($this->sut)();
        $this->assertTrue($response->terminated());
        Approvals::verifyHtml($response->output());
    }

    public function testNewHighscore(): void
    {
        $_GET = ["tetris_action" => "new_highscore"];
        $_POST = ["name" => "cmb", "score" => "10000"];
        $response = ($this->sut)();
        $this->assertTrue($response->terminated());
        $this->assertEquals([["cmb", 10000]], $this->highscoreService->readHighscores());
    }
}
