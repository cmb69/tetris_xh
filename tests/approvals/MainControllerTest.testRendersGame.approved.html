<!-- tetris -->
<div id="tetris-no-js" class="cmsimplecore_warning">Javascript must be enabled in your browser to play!</div>
<script src="./plugins/tetris/tetris.js"></script>
<div id="tetris-tabs" data-config='{"getHighscoreUrl":"/?Tetris&amp;tetris_action=get_highscore","newHighscoreUrl":"/?Tetris&amp;tetris_action=new_highscore","falldown":true,"initialSpeed":600,"acceleration":50,"labelStart":"Start","labelPause":"Pause","labelResume":"Resume"}'>
  <ul>
    <li><button id="tetris_button_play" type="button">Play</button></li>
    <li><button id="tetris_button_highscores" type="button" disabled>Highscores</button></li>
    <li><button id="tetris_button_rules" type="button">Rules</button></li>
  </ul>
  <div id="tetris" style="display:none">
    <div id="tetris-grid">
      <table>
        <tr>
          <td id="tetris-d1"></td>
          <td id="tetris-d2"></td>
          <td id="tetris-d3"></td>
          <td id="tetris-d4"></td>
          <td id="tetris-d5"></td>
          <td id="tetris-d6"></td>
          <td id="tetris-d7"></td>
          <td id="tetris-d8"></td>
          <td id="tetris-d9"></td>
          <td id="tetris-d10"></td>
        </tr>
        <tr>
          <td id="tetris-e1"></td>
          <td id="tetris-e2"></td>
          <td id="tetris-e3"></td>
          <td id="tetris-e4"></td>
          <td id="tetris-e5"></td>
          <td id="tetris-e6"></td>
          <td id="tetris-e7"></td>
          <td id="tetris-e8"></td>
          <td id="tetris-e9"></td>
          <td id="tetris-e10"></td>
        </tr>
        <tr>
          <td id="tetris-f1"></td>
          <td id="tetris-f2"></td>
          <td id="tetris-f3"></td>
          <td id="tetris-f4"></td>
          <td id="tetris-f5"></td>
          <td id="tetris-f6"></td>
          <td id="tetris-f7"></td>
          <td id="tetris-f8"></td>
          <td id="tetris-f9"></td>
          <td id="tetris-f10"></td>
        </tr>
        <tr>
          <td id="tetris-g1"></td>
          <td id="tetris-g2"></td>
          <td id="tetris-g3"></td>
          <td id="tetris-g4"></td>
          <td id="tetris-g5"></td>
          <td id="tetris-g6"></td>
          <td id="tetris-g7"></td>
          <td id="tetris-g8"></td>
          <td id="tetris-g9"></td>
          <td id="tetris-g10"></td>
        </tr>
        <tr>
          <td id="tetris-h1"></td>
          <td id="tetris-h2"></td>
          <td id="tetris-h3"></td>
          <td id="tetris-h4"></td>
          <td id="tetris-h5"></td>
          <td id="tetris-h6"></td>
          <td id="tetris-h7"></td>
          <td id="tetris-h8"></td>
          <td id="tetris-h9"></td>
          <td id="tetris-h10"></td>
        </tr>
        <tr>
          <td id="tetris-i1"></td>
          <td id="tetris-i2"></td>
          <td id="tetris-i3"></td>
          <td id="tetris-i4"></td>
          <td id="tetris-i5"></td>
          <td id="tetris-i6"></td>
          <td id="tetris-i7"></td>
          <td id="tetris-i8"></td>
          <td id="tetris-i9"></td>
          <td id="tetris-i10"></td>
        </tr>
        <tr>
          <td id="tetris-j1"></td>
          <td id="tetris-j2"></td>
          <td id="tetris-j3"></td>
          <td id="tetris-j4"></td>
          <td id="tetris-j5"></td>
          <td id="tetris-j6"></td>
          <td id="tetris-j7"></td>
          <td id="tetris-j8"></td>
          <td id="tetris-j9"></td>
          <td id="tetris-j10"></td>
        </tr>
        <tr>
          <td id="tetris-k1"></td>
          <td id="tetris-k2"></td>
          <td id="tetris-k3"></td>
          <td id="tetris-k4"></td>
          <td id="tetris-k5"></td>
          <td id="tetris-k6"></td>
          <td id="tetris-k7"></td>
          <td id="tetris-k8"></td>
          <td id="tetris-k9"></td>
          <td id="tetris-k10"></td>
        </tr>
        <tr>
          <td id="tetris-l1"></td>
          <td id="tetris-l2"></td>
          <td id="tetris-l3"></td>
          <td id="tetris-l4"></td>
          <td id="tetris-l5"></td>
          <td id="tetris-l6"></td>
          <td id="tetris-l7"></td>
          <td id="tetris-l8"></td>
          <td id="tetris-l9"></td>
          <td id="tetris-l10"></td>
        </tr>
        <tr>
          <td id="tetris-m1"></td>
          <td id="tetris-m2"></td>
          <td id="tetris-m3"></td>
          <td id="tetris-m4"></td>
          <td id="tetris-m5"></td>
          <td id="tetris-m6"></td>
          <td id="tetris-m7"></td>
          <td id="tetris-m8"></td>
          <td id="tetris-m9"></td>
          <td id="tetris-m10"></td>
        </tr>
        <tr>
          <td id="tetris-n1"></td>
          <td id="tetris-n2"></td>
          <td id="tetris-n3"></td>
          <td id="tetris-n4"></td>
          <td id="tetris-n5"></td>
          <td id="tetris-n6"></td>
          <td id="tetris-n7"></td>
          <td id="tetris-n8"></td>
          <td id="tetris-n9"></td>
          <td id="tetris-n10"></td>
        </tr>
        <tr>
          <td id="tetris-o1"></td>
          <td id="tetris-o2"></td>
          <td id="tetris-o3"></td>
          <td id="tetris-o4"></td>
          <td id="tetris-o5"></td>
          <td id="tetris-o6"></td>
          <td id="tetris-o7"></td>
          <td id="tetris-o8"></td>
          <td id="tetris-o9"></td>
          <td id="tetris-o10"></td>
        </tr>
        <tr>
          <td id="tetris-p1"></td>
          <td id="tetris-p2"></td>
          <td id="tetris-p3"></td>
          <td id="tetris-p4"></td>
          <td id="tetris-p5"></td>
          <td id="tetris-p6"></td>
          <td id="tetris-p7"></td>
          <td id="tetris-p8"></td>
          <td id="tetris-p9"></td>
          <td id="tetris-p10"></td>
        </tr>
        <tr>
          <td id="tetris-q1"></td>
          <td id="tetris-q2"></td>
          <td id="tetris-q3"></td>
          <td id="tetris-q4"></td>
          <td id="tetris-q5"></td>
          <td id="tetris-q6"></td>
          <td id="tetris-q7"></td>
          <td id="tetris-q8"></td>
          <td id="tetris-q9"></td>
          <td id="tetris-q10"></td>
        </tr>
        <tr>
          <td id="tetris-r1"></td>
          <td id="tetris-r2"></td>
          <td id="tetris-r3"></td>
          <td id="tetris-r4"></td>
          <td id="tetris-r5"></td>
          <td id="tetris-r6"></td>
          <td id="tetris-r7"></td>
          <td id="tetris-r8"></td>
          <td id="tetris-r9"></td>
          <td id="tetris-r10"></td>
        </tr>
        <tr>
          <td id="tetris-s1"></td>
          <td id="tetris-s2"></td>
          <td id="tetris-s3"></td>
          <td id="tetris-s4"></td>
          <td id="tetris-s5"></td>
          <td id="tetris-s6"></td>
          <td id="tetris-s7"></td>
          <td id="tetris-s8"></td>
          <td id="tetris-s9"></td>
          <td id="tetris-s10"></td>
        </tr>
        <tr>
          <td id="tetris-t1"></td>
          <td id="tetris-t2"></td>
          <td id="tetris-t3"></td>
          <td id="tetris-t4"></td>
          <td id="tetris-t5"></td>
          <td id="tetris-t6"></td>
          <td id="tetris-t7"></td>
          <td id="tetris-t8"></td>
          <td id="tetris-t9"></td>
          <td id="tetris-t10"></td>
        </tr>
        <tr>
          <td id="tetris-u1"></td>
          <td id="tetris-u2"></td>
          <td id="tetris-u3"></td>
          <td id="tetris-u4"></td>
          <td id="tetris-u5"></td>
          <td id="tetris-u6"></td>
          <td id="tetris-u7"></td>
          <td id="tetris-u8"></td>
          <td id="tetris-u9"></td>
          <td id="tetris-u10"></td>
        </tr>
      </table>
    </div>
    <div style="float:left">
      <div id="tetris-next">
        <table>
          <tr>
            <td id="tetris-x00"></td>
            <td id="tetris-x10"></td>
            <td id="tetris-x20"></td>
            <td id="tetris-x30"></td>
          </tr>
          <tr>
            <td id="tetris-x01"></td>
            <td id="tetris-x11"></td>
            <td id="tetris-x21"></td>
            <td id="tetris-x31"></td>
          </tr>
          <tr>
            <td id="tetris-x02"></td>
            <td id="tetris-x12"></td>
            <td id="tetris-x22"></td>
            <td id="tetris-x32"></td>
          </tr>
          <tr>
            <td id="tetris-x03"></td>
            <td id="tetris-x13"></td>
            <td id="tetris-x23"></td>
            <td id="tetris-x33"></td>
          </tr>
        </table>
      </div>
      <div id="tetris-stats">
        <div class="label">Level</div>
        <div id="tetris-level" class="led">000001</div>
        <div class="label">Rows</div>
        <div id="tetris-lines" class="led">000000</div>
        <div class="label">Score</div>
        <div id="tetris-score" class="led">000000</div>
      </div>
    </div>
    <div style="clear:both"></div>
    <div id="tetris-cmd">
      <button id="tetris-start">Start</button>
      <button id="tetris-stop" disabled="disabled">Stop</button>
    </div>
  </div>
  <div id="tetris-highscores">
    <table>
    </table>
  </div>
  <div id="tetris-rules" style="display:none">
    <div>See <a href="https://en.wikipedia.org/wiki/Tetris#Gameplay">Wikipedia</a>.</div>
    <p>Use the following keys (the arrow keys do not work in all browsers):</p>
    <table>
      <tr><td>Left</td><td class="key">J / ←</td></tr>
      <tr><td>Right</td><td class="key">L / →</td></tr>
      <tr><td>Rotate</td><td class="key">I / ↑</td></tr>
      <tr><td>Down</td><td class="key">K / ↓</td></tr>
    </table>
  </div>
</div>
